src/vorbis_decoder.cc

```diff
@@ -0,0 +1,287 @@
+ * read the actual PCM stream parameters.
+ */
+int VorbisDecoder::process_metadata(Encoder* encoder) {
+    vorbis_comment *vc = NULL;
+
+    if ((vi = ov_info(&vf, -1)) == NULL) {
+        mp3fs_debug("Ogg Vorbis decoder: Failed to retrieve the file info.");
+        return -1;
+    }
+
+    if (vi->channels > 2) {
+        mp3fs_debug("Ogg Vorbis decoder: Only mono/stereo audio currently supported.");
+        return -1;
+    }
+    
+    if (encoder->set_stream_params(
+            ov_pcm_total(&vf, -1),
+            vi->rate,
+            vi->channels) == -1) {
+        mp3fs_debug("Ogg Vorbis decoder: Failed to set encoder stream parameters.");
+        return -1;
+    }
+
+    if ((vc = ov_comment(&vf, -1)) == NULL) {
+        mp3fs_debug("Ogg Vorbis decoder: Failed to retrieve the Ogg Vorbis comment.");
+        return -1;
+    }
+        
+    double filegainref = 89.0;
+    double dbgain = INVALID_DB_GAIN;
+
+    for (int i = 0; i < vc->comments; ++i) {
+
+        /*
+         * Get the tagname - tagvalue pairs
+         */
+        std::string comment(vc->user_comments[i], vc->comment_lengths[i]);
+        unsigned long delimiter_pos = comment.find_first_of('=');
+
+        if ((delimiter_pos == 0) || (delimiter_pos >= comment.length() - 1)) {
+            continue;
+        }
+        
+        std::string tagname = comment.substr(0, delimiter_pos);
+        std::string tagvalue = comment.substr(
+                delimiter_pos + 1, comment.length() - delimiter_pos);
+
+        /*
+         * Normalize tag name to uppercase.
+         */
+        std::transform(tagname.begin(), tagname.end(),
+                tagname.begin(), ::toupper);
+
+        /*
+         * Set the encoder's text tag if it's in the metatag_map, or else,
+         * prepare the ReplayGain.
+         */
+        meta_map_t::const_iterator it = metatag_map.find(tagname);
+        
+        if (it != metatag_map.end()) {
+            encoder->set_text_tag(it->second, tagvalue.c_str());
+        }
+        else if (tagname == "REPLAYGAIN_REFERENCE_LOUDNESS") {
+            filegainref = atof(tagvalue.c_str());
+        }
+        else if (params.gainmode == 1
+                && tagname == "REPLAYGAIN_ALBUM_GAIN") {
+            dbgain = atof(tagvalue.c_str());
+        }
+        else if ((params.gainmode == 1 || params.gainmode == 2)
+                && dbgain == INVALID_DB_GAIN
+                && tagname == "REPLAYGAIN_TRACK_GAIN") {
+            dbgain = atof(tagvalue.c_str());
+        }
+    }
+
+    /*
+     * Use the Replay Gain tag to set volume scaling. The appropriate
+     * value for dbgain is set in the above if statements according to
+     * the value of gainmode. Obey the gainref option here.
+     */
+    if (dbgain != INVALID_DB_GAIN) {
+        encoder->set_gain_db(params.gainref - filegainref + dbgain);
+    }
+
+    /*
+     * TODO: decode METADATA_BLOCK_PICTURE for use with
```

Comment:
Can you do this TODO? It should be pretty straightforward if the libvorbis library has a way to access the picture structure.